
# Week 01 Assignment  
## Data Quality, Evaluation, Scaling, and Encoding

**Student name:** Tanvir Islam   

This is a small assignment that connects topics from Module 1, 2, and 3.  
You must complete it in this Colab notebook.

You will need to use concepts that appeared in the videos:
- Module 1 and 2: basic descriptive statistics, proportions, confusion matrix, accuracy, precision, recall
- Module 3: standardization, min max scaling, nominal vs ordinal, one hot encoding, ordinal encoding, Euclidean and Manhattan distance

Please do not use any extra libraries beyond `pandas`, `numpy`.



---
## 0. Setup and Dataset

We will use a dataset that should have columns given below:

- `user_id`  
- `age`  
- `monthly_income` (numeric)  
- `daily_screen_time_min` (numeric)  
- `daily_app_opens` (numeric)  
- `true_label` and `pred_label` for a binary classification task (0 or 1)  
- `satisfaction_level` (for example: `Low`, `Medium`, `High`)  
- `city_type` (for example: `Urban`, `Suburban`, `Rural`)


In [ ]:
# Cell 1: Imports
import pandas as pd
import numpy as np

In [ ]:
# Cell 2: Load the dataset (Already done for you)
df = pd.read_csv("https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220")

# Show first few rows
df.head()

user_id  age  monthly_income  daily_screen_time_min  daily_app_opens  \
0        1   43         3734.19                    109               48   
1        2   49         2594.19                    194                7   
2        3   19         3550.47                    146               36   
3        4   19         3821.18                    287               14   
4        5   63         1750.84                     66               46   

   true_label  pred_label satisfaction_level city_type  
0           0           0             Medium  Suburban  
1           0           0                Low     Urban  
2           1           0               High     Rural  
3           1           0               High  Suburban  
4           0           0             Medium  Suburban


### 0.1 Check your dataset

1. Confirm that the dataset loaded correctly.  
2. Check that you have at least these columns:  
   - numeric: `age`, `monthly_income`, `daily_screen_time_min`, `daily_app_opens`  
   - labels: `true_label`, `pred_label`  
   - categorical: `satisfaction_level`, `city_type`  



---
## Part A - Module 1 and 2 Review

In this part you will do simple descriptive statistics and basic classification evaluation.



### Q1. Descriptive statistics on a numeric feature

Choose one numeric column, for example `daily_screen_time_min`.


In [ ]:
# Q1.1: Choose your numeric column here [We already write this ans]
num_col = "daily_screen_time_min"

df[num_col].describe()

count    100.000000
mean     181.890000
std       68.886951
min       60.000000
25%      122.000000
50%      178.000000
75%      243.750000
max      299.000000
Name: daily_screen_time_min, dtype: float64


> **Q1.2 Short answer: [Marks: 05]**  
> Look at the count, mean, min, max, and standard deviation for your chosen column.  
> In 2 to 3 sentences, comment on what you see.  
> For example, does the max look very far from the mean, or does it look quite close?

Write your answer here:

>  
>  
>  



The mean (181.89 min)​ is close to the median (178 min), a roughly symmetric distribution.

The max (299 min)​ is much higher than the mean

The standard deviation (68.88 min)​ is relatively large, meaning there is significant variability in daily screen time.




### Q2. Proportion of positive class

Use the `true_label` column, where 1 means "positive" and 0 means "negative".


In [ ]:
# Q2.1: Compute proportion of positive class [We already write this ans]
label_col = "true_label"

positive_count = (df[label_col] == 1).sum()
total_count = df.shape[0]
positive_proportion = positive_count / total_count

print("Positive count:", positive_count)
print("Total samples:", total_count)
print("Proportion of positive class:", positive_proportion)

Positive count: 52
Total samples: 100
Proportion of positive class: 0.52



> **Q2.2 Short answer: [5 marks]**  
> In 1 to 2 sentences, explain what this proportion tells you about your dataset.  
> For example, is the dataset balanced between 0 and 1, or is one class much more common?

Write your answer here:

>  
>  
>  


The dataset is slightly imbalanced​ (52% positive, 48% negative).

It is almost balanced, but not perfectly (50-50 split).


### Q3. Confusion matrix and basic metrics

For this question, use:
- `true_label` as the actual label  
- `pred_label` as the model prediction


In [ ]:
# Q3.1: Manually compute TP, TN, FP, FN [We already write this ans]
true_col = "true_label"
pred_col = "pred_label"

tp = ((df[true_col] == 1) & (df[pred_col] == 1)).sum()
tn = ((df[true_col] == 0) & (df[pred_col] == 0)).sum()
fp = ((df[true_col] == 0) & (df[pred_col] == 1)).sum()
fn = ((df[true_col] == 1) & (df[pred_col] == 0)).sum()

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 28
TN: 27
FP: 21
FN: 24


In [ ]:
# Q3.2: Compute accuracy, precision, recall [We already write this ans]
accuracy = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0

precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)

Accuracy: 0.55
Precision: 0.5714285714285714
Recall: 0.5384615384615384



> **Q3.3 Short answer: [10 marks]**  
> In 3 to 4 sentences, briefly comment on the model using these three metrics.  
> For example, is the model catching most positives (high recall) or being careful when it predicts positive (high precision)?

Write your answer here:

>  
>  
>  


Accuracy (0.55)​ is moderate, meaning the model is slightly better than random guessing.

Precision (0.571)​ is decent, meaning when the model predicts 1 (positive), it is correct 57%​ of the time.

Recall (0.538)​ is also moderate, meaning the model captures 54%​ of actual positive cases.


---
## Part B - Module 3: Scaling and Encoding

Now we will pick a few features and apply scaling and encoding.



### Q4. Standardization and Min max scaling

Use one numeric column, `monthly_income`.


In [ ]:
# Q4.1: Choose the numeric column [2 marks]z

In [ ]:
income = df["monthly_income"]

In [ ]:
# Q4.2: Standardization with z-score [10 marks]

In [ ]:

mean = income.mean()
std = income.std()
income_std = (income - mean) / std
print("Standardized (Z-Score):\n", income_std.head())

Standardized (Z-Score):
 0    0.944685
1   -0.324626
2    0.740126
3    1.041542
4   -1.263639
Name: monthly_income, dtype: float64


In [ ]:
# Q4.3: Min max scaling implementation [10 marks]

In [ ]:
min_val = income.min()
max_val = income.max()
income_minmax = (income - min_val) / (max_val - min_val)
print("Min-Max Scaled:\n", income_minmax.head())

Min-Max Scaled:
 0    0.675209
1    0.393685
2    0.629839
3    0.696691
4    0.185420
Name: monthly_income, dtype: float64



> **Q4.4 Short answer: [3 marks]**  
> Compare the standardized and min max scaled columns in 2 to 3 sentences.  
> Mention what kind of range each one uses and how the numbers look.

Write your answer here:

>  
>  
>  


Standardized (Z-Score):​

Range:​ Unbounded ( negative or >1)

Mean = 0, Std = 1​

Useful for algorithms sensitive to outliers.

Min-Max Scaled:​

Range:​ [0, 1]

All values are normalized between 0 and 1.

Useful for  when you need a fixed range.


### Q5. One hot and ordinal encoding

We will use:
- `city_type` as a nominal feature  
- `satisfaction_level` as an ordinal feature with order `Low` < `Medium` < `High`  


In [ ]:
# Q5.1: One hot encoding for city_type using pandas [10 marks]

In [ ]:
city_onehot = pd.get_dummies(df["city_type"], prefix="city")
print(city_onehot.head())

   city_Rural  city_Suburban  city_Urban
0       False           True       False
1       False          False        True
2        True          False       False
3       False           True       False
4       False           True       False


In [ ]:
# Q5.2: Attach one hot encoded columns to df [5 marks]

In [ ]:
df = pd.concat([df, city_onehot], axis=1)

In [ ]:
# Q5.3: Ordinal encoding for satisfaction_level [10 marks]

In [ ]:
satisfaction_map = {"Low": 0, "Medium": 1, "High": 2}
df["satisfaction_encoded"] = df["satisfaction_level"].map(satisfaction_map)
print(df[["satisfaction_level", "satisfaction_encoded"]].head())

  satisfaction_level  satisfaction_encoded
0             Medium                     1
1                Low                     0
2               High                     2
3               High                     2
4             Medium                     1



> **Q5.4 Short answer: [5 marks]**  
> In 2 to 3 sentences, explain why one hot encoding is suitable for `city_type`  
> and why ordinal encoding is suitable for `satisfaction_level`.

Write your answer here:

>  
>  
>  


One-hot encoding is suitable for city_type because it is a nominal feature with no inherent order

Ordinal encoding is suitable for satisfaction_level because it is an ordinal feature with a clear ranking: Low < Medium < High


---
## Part C - Module 3: Distances between users

For this small part we will work with vectors based on scaled numeric features.



### Q6. Euclidean and Manhattan distance

Build 2D vectors for user 0 and user 1 using:
- `income_std`  
- `daily_app_opens` (or its min max scaled version if you prefer)


In [ ]:
# Q6.1: Build 2D vectors for first two users [We already write this ans]
vec_cols = ["monthly_income", "daily_app_opens"]

v1 = df.loc[0, vec_cols].values
v2 = df.loc[1, vec_cols].values

print("v1:", v1)
print("v2:", v2)

v1: [np.float64(3734.19) np.int64(48)]
v2: [np.float64(2594.19) np.int64(7)]


In [ ]:
# Q6.2: Euclidean distance computation [10 marks]

In [ ]:

euclidean = np.sqrt(np.sum((v1 - v2)**2))
print("Euclidean distance:", euclidean)

Euclidean distance: 1140.7370424422975


In [ ]:
# Q6.3: Manhattan distance computation [10 marks]

In [ ]:
# Q6.3 Manhattan distance
manhattan = np.sum(np.abs(v1 - v2))
print("Manhattan distance:", manhattan)

Manhattan distance: 1181.0



> **Q6.4 Short answer: [5 marks]**  
> Which one is larger in your result, Euclidean or Manhattan distance  
> and why does that usually happen based on their formulas?

Write your answer here:

>  
>  
>  


Manhattan (1181.0) > Euclidean (1140.12)

​The Manhattan distance (1181.0) is larger than the Euclidean distance (1140.12).
The Manhattan distance is the sum of absolute differences.
and the Euclidean distance is the square root of the sum of squared differences.
so Manhattan is always greater than or equal to Euclidean


---
## Final Reflection [5 marks]

> In 4 to 6 sentences, describe how the three modules connect in this assignment.  
> Mention:
> - One idea from Module 1 or 2 that you used  
> - One idea from Module 3 that you used  
> - How these ideas together help you understand a dataset more deeply

Write your reflection here:

>  
>  
>  


Module 1/2 (Descriptive Stats & Classification):​

Used mean, std, confusion matrix, accuracy, precision, recall​ to understand data distribution and model performance.

Module 3 (Scaling & Encoding):​

Applied standardization, min-max scaling, one-hot & ordinal encoding​ to prepare data for ML models.

Combined Insight:​

Descriptive stats​ helped identify data spread, classification metrics​ evaluated model quality, and scaling/encoding​ ensured data is ML-ready.




## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Download this notebook as `.ipynb` and upload it according to the given instructions.
- ***Must Read Assignment Module Text Instruction fully Where you will find how to submit this assignment***
